packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2755,7 +2755,9 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       renderEditable.setFloatingCursor(FloatingCursorDragState.End, finalPosition, _lastTextPosition!);
       // Only change if new position is out of current selection range, as the
       // selection may have been modified using the iOS keyboard selection gesture.
-      if (_lastTextPosition!.offset < renderEditable.selection!.start || _lastTextPosition!.offset >= renderEditable.selection!.end) {
+      if (renderEditable.selection!.isCollapsed
+          || _lastTextPosition!.offset < renderEditable.selection!.start
```

Comment:
Should affinity be taken into consideration?

Example: if `renderEditable.selection.start` points to the end position of a word wrap, and `_lastTextPosition` points to the start position of a word wrap, then we probably want to call `_handleSelectionChanged`?
```
abc <- _lastTextPosition
def
^
 -------- renderEditable.selection.start
```

Reply:
There isn't any way to get those variables in the state you show, since if you move the floating cursor up to the first line, UIKit will update the selection to include that line.

I think actually the new check of `renderEditable.selection!.isCollapsed` covers everything.

- If true, keyboard selection never ran, so we should call _handleSelectionChanged
- If false, then keyboard selection certainly ran and UIKit has set the selection, so we shouldn't change it

Since there isn't any other way to set a non-collapsed selection during floating cursor besides the keyboard gesture.

Reply:
Does a single tap force touch collapse the current selection as well? IIUC double tap force touches trigger selection changes in the text input plugin so the framework doesn't have to handle that. And this part of framework code is to adjust the selection when a single tap force touch completes, since UIKit only changes the selection for double tap force touches?

Reply:
Yep, the selection is collapsed at the beginning of keyboard gesture, so it applies to both single and double tap. 

This code here gets called when either single or double tap completes. So we need to make sure to only change the selection if it's a single tap.

Reply:
Thanks for the clarification!